src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/MsaPassthroughHandler.cs

```diff
@@ -0,0 +1,146 @@
+        private readonly IWamProxy _wamProxy;
+        private readonly IntPtr _parentHandle;
+
+        public MsaPassthroughHandler(
+            ICoreLogger logger,
+            IWamPlugin msaPlugin,
+            IWamProxy wamProxy,
+            IntPtr parentHandle)
+        {
+            _logger = logger;
+            _msaPlugin = msaPlugin;
+            _wamProxy = wamProxy;
+            _parentHandle = parentHandle;
+        }
+
+        public bool IsPassthroughEnabled(AuthenticationRequestParameters authenticationRequestParameters)
+        {
+            // TODO: change this to a config object
+            bool enabled = authenticationRequestParameters.ExtraQueryParameters.TryGetValue("msal_msa_pt", out string val) &&
+                string.Equals("1", val);
+            _logger.Info("WAM MSA-PT: PassThrough is enabled? " + enabled);
+
+            return enabled;
+        }
+
+        public async Task<string> FetchTransferTokenAsync(AuthenticationRequestParameters authenticationRequestParameters, WebAccountProvider accountProvider)
+        {
+            _logger.Verbose("WAM MSA-PT - fetching transfer token");
+            string transferToken = await FetchMsaPassthroughTransferTokenAsync(
+                authenticationRequestParameters, accountProvider)
+                .ConfigureAwait(false);
+
+            return transferToken;
+        }
+
+        public void AddTransferTokenToRequest(
+            Windows.Security.Authentication.Web.Core.WebTokenRequest webTokenRequest,
+            string transferToken)
+        {
+            if (!string.IsNullOrEmpty(transferToken))
+            {
+                webTokenRequest.Properties.Add("SamlAssertion", transferToken);
+                webTokenRequest.Properties.Add("SamlAssertionType", "SAMLV1");
+            }
+        }
+
+        private async Task<string> FetchMsaPassthroughTransferTokenAsync(
+           AuthenticationRequestParameters authenticationRequestParameters,
+           WebAccountProvider accountProvider)
+        {
+            // step 1 - get a response from the MSA provider, just to have a WebAccount
+            _logger.Info("WAM MSA-PT: Making initial call to MSA provider");
+            WebAccount msaPtWebAccount = await FetchWebAccountFromMsaAsync(
+                authenticationRequestParameters,
+                accountProvider).ConfigureAwait(false);
+
+            // step 2 - get a trasnfer token 
+            _logger.Info("WAM MSA-PT: Getting transfer token");
+            string transferToken = await FetchTransferTokenAsync(
+                accountProvider,
+                msaPtWebAccount,
+                authenticationRequestParameters.ClientId).ConfigureAwait(true);
+
+            return transferToken;
+        }
+
+        private async Task<WebAccount> FetchWebAccountFromMsaAsync(
+            AuthenticationRequestParameters authenticationRequestParameters, WebAccountProvider accountProvider)
+        {
```

Comment:
This is the magic for MSA-PT: 

- get a v1 token from MSA plugin
- use that to get a SAML transfer token
- request a v2 token from the AAD plugin

Reply:
this seems like a big perf issue. is WAM not able to do this? seems like something WAM should do.

---
In reply to: [564741608](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2365#discussion_r564741608) [](ancestors = 564741608)

Reply:
Well, this is interactive auth, so perf is not a big issue. 
Implementation is based on C++ existing impl and advice.